File: sdk/python/aragora_sdk/namespaces/notifications.py
"""
Notifications Namespace API

Provides endpoints for managing notification preferences and sending notifications.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class NotificationsAPI:
    """Synchronous Notifications API."""

    def __init__(self, client: AragoraClient):
        self._client = client

    def send(self, **kwargs: Any) -> dict[str, Any]:
        """Send a notification."""
        return self._client.request("POST", "/api/v1/notifications/send", json=kwargs)

    def get_status(self) -> dict[str, Any]:
        """Get notification service status."""
        return self._client.request("GET", "/api/v1/notifications/status")

    def get_history(self) -> dict[str, Any]:
        """Get notification history."""
        return self._client.request("GET", "/api/v1/notifications/history")

    def configure_email(self, **kwargs: Any) -> dict[str, Any]:
        """Configure email notifications."""
        return self._client.request("POST", "/api/v1/notifications/email/config", json=kwargs)

    def list_email_recipients(self) -> dict[str, Any]:
        """List email notification recipients."""
        return self._client.request("GET", "/api/v1/notifications/email/recipients")

    def add_email_recipient(self, **kwargs: Any) -> dict[str, Any]:
        """Add an email notification recipient."""
        return self._client.request("POST", "/api/v1/notifications/email/recipient", json=kwargs)

    def remove_email_recipient(self, **kwargs: Any) -> dict[str, Any]:
        """Remove an email notification recipient."""
        return self._client.request("DELETE", "/api/v1/notifications/email/recipient", json=kwargs)

    def configure_telegram(self, **kwargs: Any) -> dict[str, Any]:
        """Configure Telegram notifications."""
        return self._client.request("POST", "/api/v1/notifications/telegram/config", json=kwargs)

    def test(self, **kwargs: Any) -> dict[str, Any]:
        """Send a test notification."""
        return self._client.request("POST", "/api/v1/notifications/test", json=kwargs)

    def get_delivery_stats(self) -> dict[str, Any]:
        """Get notification delivery statistics (success rate, latency, failures)."""
        return self._client.request("GET", "/api/notifications/delivery-stats")

    def get_preferences(self) -> dict[str, Any]:
        """Get notification preferences for the current user."""
        return self._client.request("GET", "/api/notifications/preferences")

    def update_preferences(self, **kwargs: Any) -> dict[str, Any]:
        """
        Update notification preferences.

        Args:
            **kwargs: Preferences (channels, frequency, quiet_hours, etc.)

        Returns:
            Dict with updated preferences.
        """
        return self._client.request("PUT", "/api/notifications/preferences", json=kwargs)

    def list_templates(self) -> dict[str, Any]:
        """List notification templates for the current user."""
        return self._client.request("GET", "/api/notifications/templates")

    def get_template(self, template_id: str) -> dict[str, Any]:
        """Get a notification template by ID."""
        return self._client.request("GET", f"/api/notifications/templates/{template_id}")

    def update_template(
        self,
        template_id: str,
        *,
        subject: str | None = None,
        body: str | None = None,
    ) -> dict[str, Any]:
        """Update subject/body overrides for a notification template."""
        payload: dict[str, Any] = {}
        if subject is not None:
            payload["subject"] = subject
        if body is not None:
            payload["body"] = body
        return self._client.request(
            "PUT", f"/api/notifications/templates/{template_id}", json=payload
        )

    def reset_template(self, template_id: str) -> dict[str, Any]:
        """Reset a notification template to its default content."""
        return self._client.request("POST", f"/api/notifications/templates/{template_id}/reset")

    def preview_template(
        self,
        template_id: str,
        *,
        values: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Render a notification template with preview values."""
        payload = {"values": values} if values is not None else {}
        return self._client.request(
            "POST",
            f"/api/notifications/templates/{template_id}/preview",
            json=payload,
        )


class AsyncNotificationsAPI:
    """Asynchronous Notifications API."""

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def send(self, **kwargs: Any) -> dict[str, Any]:
        """Send a notification."""
        return await self._client.request("POST", "/api/v1/notifications/send", json=kwargs)

    async def get_status(self) -> dict[str, Any]:
        """Get notification service status."""
        return await self._client.request("GET", "/api/v1/notifications/status")

    async def get_history(self) -> dict[str, Any]:
        """Get notification history."""
        return await self._client.request("GET", "/api/v1/notifications/history")

    async def configure_email(self, **kwargs: Any) -> dict[str, Any]:
        """Configure email notifications."""
        return await self._client.request("POST", "/api/v1/notifications/email/config", json=kwargs)

    async def list_email_recipients(self) -> dict[str, Any]:
        """List email notification recipients."""
        return await self._client.request("GET", "/api/v1/notifications/email/recipients")

    async def add_email_recipient(self, **kwargs: Any) -> dict[str, Any]:
        """Add an email notification recipient."""
        return await self._client.request(
            "POST", "/api/v1/notifications/email/recipient", json=kwargs
        )

    async def remove_email_recipient(self, **kwargs: Any) -> dict[str, Any]:
        """Remove an email notification recipient."""
        return await self._client.request(
            "DELETE", "/api/v1/notifications/email/recipient", json=kwargs
        )

    async def configure_telegram(self, **kwargs: Any) -> dict[str, Any]:
        """Configure Telegram notifications."""
        return await self._client.request(
            "POST", "/api/v1/notifications/telegram/config", json=kwargs
        )

    async def test(self, **kwargs: Any) -> dict[str, Any]:
        """Send a test notification."""
        return await self._client.request("POST", "/api/v1/notifications/test", json=kwargs)

    async def get_delivery_stats(self) -> dict[str, Any]:
        """Get notification delivery statistics."""
        return await self._client.request("GET", "/api/notifications/delivery-stats")

    async def get_preferences(self) -> dict[str, Any]:
        """Get notification preferences for the current user."""
        return await self._client.request("GET", "/api/notifications/preferences")

    async def update_preferences(self, **kwargs: Any) -> dict[str, Any]:
        """Update notification preferences."""
        return await self._client.request("PUT", "/api/notifications/preferences", json=kwargs)

    async def list_templates(self) -> dict[str, Any]:
        """List notification templates for the current user."""
        return await self._client.request("GET", "/api/notifications/templates")

    async def get_template(self, template_id: str) -> dict[str, Any]:
        """Get a notification template by ID."""
        return await self._client.request("GET", f"/api/notifications/templates/{template_id}")

    async def update_template(
        self,
        template_id: str,
        *,
        subject: str | None = None,
        body: str | None = None,
    ) -> dict[str, Any]:
        """Update subject/body overrides for a notification template."""
        payload: dict[str, Any] = {}
        if subject is not None:
            payload["subject"] = subject
        if body is not None:
            payload["body"] = body
        return await self._client.request(
            "PUT",
            f"/api/notifications/templates/{template_id}",
            json=payload,
        )

    async def reset_template(self, template_id: str) -> dict[str, Any]:
        """Reset a notification template to its default content."""
        return await self._client.request(
            "POST", f"/api/notifications/templates/{template_id}/reset"
        )

    async def preview_template(
        self,
        template_id: str,
        *,
        values: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Render a notification template with preview values."""
        payload = {"values": values} if values is not None else {}
        return await self._client.request(
            "POST",
            f"/api/notifications/templates/{template_id}/preview",
            json=payload,
        )

File: sdk/typescript/src/namespaces/notifications.ts
/**
 * Notifications Namespace API
 *
 * Provides access to notification settings and delivery for email, Telegram, and webhooks.
 */

/**
 * Notification channel types
 */
export type NotificationChannel = 'email' | 'telegram' | 'webhook' | 'slack';

/**
 * Notification event types
 */
export type NotificationEventType =
  | 'debate_started'
  | 'debate_completed'
  | 'consensus_reached'
  | 'consensus_failed'
  | 'agent_error'
  | 'budget_alert'
  | 'gauntlet_completed'
  | 'security_alert';

/**
 * Integration status for a notification channel
 */
export interface IntegrationStatus {
  channel: NotificationChannel;
  enabled: boolean;
  configured: boolean;
  last_delivery?: string;
  last_error?: string;
}

/**
 * Email configuration settings
 */
export interface EmailConfig {
  smtp_host?: string;
  smtp_port?: number;
  smtp_user?: string;
  smtp_password?: string;
  from_address?: string;
  use_tls?: boolean;
}

/**
 * Telegram configuration settings
 */
export interface TelegramConfig {
  bot_token?: string;
  chat_id?: string;
  parse_mode?: 'HTML' | 'Markdown' | 'MarkdownV2';
}

/**
 * Email recipient entry
 */
export interface EmailRecipient {
  email: string;
  name?: string;
  events?: NotificationEventType[];
  added_at?: string;
}

/**
 * Notification delivery record
 */
export interface NotificationDelivery {
  id: string;
  channel: NotificationChannel;
  event_type: NotificationEventType;
  recipient: string;
  subject?: string;
  delivered_at: string;
  success: boolean;
  error?: string;
}

export interface NotificationTemplate {
  id: string;
  name: string;
  description?: string;
  channel?: NotificationChannel | string;
  subject: string;
  body: string;
  variables?: string[];
  sample_values?: Record<string, string>;
  customized?: boolean;
}

export interface NotificationTemplatePreview {
  template_id: string;
  rendered_subject: string;
  rendered_body: string;
  values_used: Record<string, unknown>;
}

/**
 * Interface for the internal client used by NotificationsAPI.
 */
interface NotificationsClientInterface {
  request<T>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Notifications API namespace.
 *
 * Provides methods for managing notification channels and delivery:
 * - Configure email, Telegram, and webhook integrations
 * - Manage recipients and subscriptions
 * - Send test notifications
 * - View delivery history
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // Check notification status
 * const status = await client.notifications.getStatus();
 *
 * // Configure email notifications
 * await client.notifications.configureEmail({
 *   smtp_host: 'smtp.example.com',
 *   from_address: 'aragora@example.com'
 * });
 *
 * // Send a test notification
 * await client.notifications.sendTest({ channel: 'email' });
 * ```
 */
export class NotificationsAPI {
  constructor(private client: NotificationsClientInterface) {}

  /**
   * Get notification integration status for all channels.
   */
  async getStatus(): Promise<{ integrations: IntegrationStatus[] }> {
    return this.client.request('GET', '/api/notifications/status');
  }

  /**
   * Configure email notification settings.
   */
  async configureEmail(config: EmailConfig): Promise<{ success: boolean; message: string }> {
    return this.client.request('POST', '/api/notifications/email/config', {
      json: config as unknown as Record<string, unknown>,
    });
  }

  /**
   * Configure Telegram notification settings.
   */
  async configureTelegram(config: TelegramConfig): Promise<{ success: boolean; message: string }> {
    return this.client.request('POST', '/api/notifications/telegram/config', {
      json: config as unknown as Record<string, unknown>,
    });
  }

  /**
   * Add an email recipient for notifications.
   */
  async addEmailRecipient(recipient: {
    email: string;
    name?: string;
    events?: NotificationEventType[];
  }): Promise<{ success: boolean; recipient: EmailRecipient }> {
    return this.client.request('POST', '/api/notifications/email/recipient', { json: recipient });
  }

  /**
   * Remove an email recipient.
   */
  async removeEmailRecipient(email: string): Promise<{ success: boolean }> {
    return this.client.request('DELETE', '/api/notifications/email/recipient', {
      json: { email },
    });
  }

  /**
   * List all email recipients.
   */
  async listEmailRecipients(): Promise<{ recipients: EmailRecipient[] }> {
    return this.client.request('GET', '/api/notifications/email/recipients');
  }

  /**
   * Send a test notification to verify configuration.
   */
  async sendTest(options: {
    channel: NotificationChannel;
    recipient?: string;
  }): Promise<{ success: boolean; message: string }> {
    return this.client.request('POST', '/api/notifications/test', { json: options });
  }

  /**
   * Send a notification immediately.
   */
  async send(notification: {
    channel: NotificationChannel;
    event_type: NotificationEventType;
    recipient?: string;
    subject?: string;
    message: string;
    metadata?: Record<string, unknown>;
  }): Promise<{ success: boolean; delivery_id?: string }> {
    return this.client.request('POST', '/api/notifications/send', { json: notification });
  }

  /**
   * Get notification delivery history.
   */
  async getDeliveryHistory(options?: {
    channel?: NotificationChannel;
    event_type?: NotificationEventType;
    success?: boolean;
    limit?: number;
    offset?: number;
  }): Promise<{ deliveries: NotificationDelivery[]; total: number }> {
    return this.client.request('GET', '/api/notifications/history', { params: options });
  }

  /**
   * Get notification delivery statistics (success rate, latency, failures).
   */
  async getDeliveryStats(): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/notifications/delivery-stats');
  }

  /**
   * Update notification preferences for the current user.
   *
   * @param preferences - Notification preferences (channels, frequency, quiet hours, etc.)
   * @returns Updated preferences
   */
  async updatePreferences(preferences: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('PUT', '/api/v1/notifications/preferences', {
      json: preferences,
    });
  }

  /**
   * List notification templates for the current user.
   */
  async listTemplates(): Promise<{ templates: NotificationTemplate[]; count: number }> {
    return this.client.request('GET', '/api/notifications/templates');
  }

  /**
   * Get a notification template by ID.
   */
  async getTemplate(templateId: string): Promise<{ template: NotificationTemplate }> {
    return this.client.request('GET', `/api/notifications/templates/${templateId}`);
  }

  /**
   * Update subject/body overrides for a notification template.
   */
  async updateTemplate(
    templateId: string,
    updates: { subject?: string; body?: string }
  ): Promise<{ template: NotificationTemplate; updated: boolean }> {
    return this.client.request('PUT', `/api/notifications/templates/${templateId}`, {
      json: updates,
    });
  }

  /**
   * Reset a notification template to its default content.
   */
  async resetTemplate(
    templateId: string
  ): Promise<{ template: NotificationTemplate; reset: boolean }> {
    return this.client.request('POST', `/api/notifications/templates/${templateId}/reset`);
  }

  /**
   * Render a notification template with preview values.
   */
  async previewTemplate(
    templateId: string,
    values?: Record<string, unknown>
  ): Promise<NotificationTemplatePreview> {
    return this.client.request('POST', `/api/notifications/templates/${templateId}/preview`, {
      json: values ? { values } : {},
    });
  }
}
